s3/replication/common/experiments/kafka_consumer.py

```diff
@@ -0,0 +1,162 @@
+        self._partition = KafkaMain.job_id_to_offset[job_id]["Partition"]
+        return self._partition
+
+    def get_offset(self, job_id):
+        self._offset = KafkaMain.job_id_to_offset[job_id]["Offset"]
+        return self._offset
+
+    # Remove job from job-id to offset mapping after commit
+    @staticmethod
+    def remove_job_from_mapping(job_id):
+        removed_key = KafkaMain.job_id_to_offset.pop(job_id, None)
+        if removed_key is not None:
+            print("Removed job id from mapping : {}".format(removed_key))
+        else:
+            print("Job id [{}] does not exist".format(job_id))
+
+    # Send job for replication after consuming from kafka
+    async def sendJob(self, data):
+        print("Sending Jobs...")
+        fdmi_dict = literal_eval(data)
+        job_record = PrepareReplicationJob.from_fdmi(fdmi_dict)
+
+        if job_record is None:
+            print("BadRequest")
+            return
+
+        jobs_list = self._app['all_jobs']
+        if jobs_list.is_job_present(job_record["replication-id"]):
+            # Duplicate job.
+            print('Duplicate Job!')
+            return
+
+        job = jobs_list.add_job_using_json(job_record)
+        self._job_id = job.get_job_id()
+        print("Added Job : {}".format(self._job_id))
+
+        # Add job id and consumer details list to dictionary
+        KafkaMain.job_id_to_offset[self._job_id] = self._consumer_details
+        print("Job-id to offset mapping : {}".format(KafkaMain.job_id_to_offset))
+        self._consumer_details = {}
+
+    # Start the consumer and add the consumer details into list
+    async def consume(self):
+        await KafkaMain.consumer.start()
+        topic_partition = TopicPartition('FDMI', 0)
+        position = await KafkaMain.consumer.position(topic_partition)
+
+        try:
+            print("In consumer loop")
+            # Consume messages
+            async for msg in KafkaMain.consumer:
+                if msg.offset == position:
+                    pass
+                else:
+                    obj = json.loads(msg.value.decode())
+                    print("Event consumed : {}", obj['cr_val'])
+
+                    self._consumer_details["Topic"] = msg.topic
+                    self._consumer_details["Partition"] = msg.partition
+                    self._consumer_details["Offset"] = msg.offset
+                    print(self._consumer_details)
+
+                    # Send jobs for replication
+                    await self.sendJob(obj['cr_val'])
+        except Exception as e:
+            print("Exception : {}".format(e))
+        finally:
+            await KafkaMain.consumer.stop()
+            # Close the kafka consumer
+            KafkaMain.consumer.close()
+
+    # Manually commit offset after replication is completed
+    # This method should be called from job_routes.py
+    # After getting replication status as completed
+    async def commit_job(self, job_id):
```

Comment:
Sorry, I am new to python async/await.

What part of this function is asynchronously?

I don't see how this "KafkaMain" is used.
And why the aiokafka is needed?

Reply:
okay.
I had tested the code with kafka-python module by integrating consumer code with replication manager. At that time, consumer was started successfully and also consumed events from kafka, but distributor couldn't start it's loop, and hence replicators couldn't connect to replication manager. I used aiokafka because, we want to start the consumer and distributor loop simultaneously when we run the replication manager.